h/auth/util.py

```diff
@@ -109,3 +115,38 @@ def authority(request):
     Falls back on returning request.domain if h.authority isn't set.
     """
     return text_type(request.registry.settings.get('h.authority', request.domain))
+
+
+def request_auth_client(request):
+    creds = basic_auth_creds(request)
+    if creds is None:
+        raise ClientUnauthorized()
+
+    # We fetch the client by its ID and then do a constant-time comparison of
+    # the secret with that provided in the request.
+    #
+    # It is important not to include the secret as part of the SQL query
+    # because the resulting code may be subject to a timing attack.
+    client_id, client_secret = creds
+    try:
+        client = request.db.query(AuthClient).get(client_id)
+    except sa.exc.StatementError:  # client_id is malformed
+        raise ClientUnauthorized()
+    if client is None:
+        raise ClientUnauthorized()
+    if client.secret is None:  # client is not confidential
+        raise ClientUnauthorized()
+    if client.grant_type != GrantType.client_credentials:  # client not allowed to create users
+        raise ClientUnauthorized()
+
+    if not hmac.compare_digest(client.secret, client_secret):
+        raise ClientUnauthorized()
+
+    return client
+
+
+def validate_auth_client_authority(client, data):
+    authority = data.get('authority')
+    if client.authority != authority:
+        msg = "'authority' does not match authenticated client"
+        raise ValidationError(msg)
```

Comment:
I think the previous line and this one can be combined into a single line (again, I know you found it like this, just might as well spiff it up while we're here!)